Remove the deprecated '--kernel-memory' option

aiordache opened the pull request:
Ignore the kernel memory option if set in `HostConfig` on container create. 
We keep the option in the API ( for the `containers/create` and `/info` endpoints) until the next API version release but ignore any value it is set to.

```
root@80b7d3e03521:~# curl --unix-socket /var/run/docker.sock -H "Content-Type: application/json" -d '{"Image": "alpine", "Cmd": ["echo", "hello world"], "HostConfig": {"KernelMemory": 299900}}' -X POST http://localhost/v1.41/containers/create
{"Id":"4d8537d9cee548938492d489b7d8394b46ae75992d89875fbf2b33971762a552","Warnings":["Specifying a kernel memory limit is not supported anymore."]}
```
```
root@80b7d3e03521:~# docker inspect 4d8537d9cee5
[
    {
        "Id": "4d8537d9cee548938492d489b7d8394b46ae75992d89875fbf2b33971762a552",
....
            "KernelMemory": 0,
            "KernelMemoryTCP": 0,
            "MemoryReservation": 0,
....

```
```
root@80b7d3e03521:~# curl -s --unix-socket /var/run/docker.sock http://localhost/info | jq .| grep KernelMemory
  "KernelMemory": false,
  "KernelMemoryTCP": false,
```

tianon commented on daemon/daemon_unix.go:
```diff
@@ -102,7 +102,7 @@ func getMemoryResources(config containertypes.Resources) *specs.LinuxMemory {
 
 	if config.KernelMemory != 0 {
-		memory.Kernel = &config.KernelMemory
+		logrus.Warn("Your kernel does not support kernel memory limit capabilities or the cgroup is not mounted. Limitation discarded.")
```
This message seems inaccurate here -- to give this warning, shouldn't it check `sysInfo.KernelMemory` first like it did previously?  Alternatively, perhaps we don't warn about kernel support at all, given we're going to ignore the container setting entirely?

tianon commented:
Just to thread the needle a little bit, this is related to https://github.com/opencontainers/runtime-spec/pull/1093 and https://github.com/opencontainers/runc/pull/2840.

thaJeztah commented:
> Just to thread the needle a little bit, this is related to [opencontainers/runtime-spec#1093](https://github.com/opencontainers/runtime-spec/pull/1093) and [opencontainers/runc#2840](https://github.com/opencontainers/runc/pull/2840).

Yes; also related to https://github.com/opencontainers/runc/issues/3174 (which makes things more complicated)


FWIW; I discussed with @aiordache to change this implementation to

- ignore the option in the API for API 1.42 and up (instead of removing it everywhere)
- remove the"your host does not support ..." warnings (which this PR also does)
- update the API documentation (swagger) for 1.42 to remove mention of the field, and to warn users that the "kernel-memory-tcp" is supported by the API but won't take effect (depending on the version of runc and/or OCI runtime being used, see below).


I had a draft branch where I started working on this, (https://github.com/moby/moby/compare/master...thaJeztah:remove_kmem_options) which is when I found out that kmem.tcp was _not_ deprecated in the kernel, so runc (and the spec to an extend) had been a bit too optimistic deprecating the options.

So now we're in a situation where;

- we deprecated the `--kernel-memory` option, but did _NOT_ (yet?) deprecate `--kernel-memory-tcp`
- runc v1.0.0 _removed_ both (so even if we would set them, they would take no effect)
- other runtimes _MAY_ still support them (as well as older versions of runc)
- the runtime spec "discourages" both

So what's the best approach here?

1. Fully remove the `--kernel-memory` option (current implementation in this PR); technically "incorrect" as we'd be modifying older API versions (which should be considered "frozen")
2. Ignore the `--kernel-memory` option for API 1.42 and up (proposed change), knowing that current runc versions will ignore it (but other runtimes may still support it, so keeping that use-case)
3. same as `1.` or `2.` but for _both_ `--kernel-memory` _and_ `--kernel-memory-tcp` (problem: we didn't announce the latter one to be deprecated)
4. do `1.` or `2.`, but _un-deprecate_ the `--kernel-memory-tcp` in runc (TBD: https://github.com/opencontainers/runc/issues/3174 - it's not 100% clear to me if re-introducing that option also brings back the problems with the broken 3.10 RHEL kernels)

Commit:
Remove KernelMemory option from /containers/create endpoint
 - remove KernelMemory option from `v1.42` api docs
 - remove KernelMemory warning on `/info`
 - update changes for `v1.42`

Signed-off-by: aiordache <anca.iordache@docker.com>

thaJeztah reviewed (changes requested):
Arf; of course it helps if I actually submit my review :joy:

Looks good; some minor nits:

- the first commit message has a typo (`s/KernelMemeory/KernelMemory/`)
- perhaps you can squash at least the first two commits (I think those should go together)
- could you add a bullet to https://github.com/moby/moby/blob/master/docs/api/version-history.md to mention the field has been removed from the API and will be ignored for API v1.42 and above?

thaJeztah commented on docs/api/version-history.md:
```diff
@@ -34,6 +34,9 @@ keywords: "API, Docker, rcli, REST, documentation"
   (`a disk usage operation is already running`) would be returned in this
   situation. This change is not versioned, and affects all API versions if the
   daemon has this patch.
+* Removed the `KernelMemory` field in `HostConfig` on the `POST /containers/create`
```
🤦🤦🤦 So I'm reading this, and just realised I had to check if `docker container update` also supported this, and it looks like it does, so we need to update the `/containers/{id}/update` endpoint as well; https://github.com/moby/moby/blob/9e7bbdb9ba53440cfc67e2e1de89e35e44c43ddf/api/server/router/container/container_routes.go#L421-L437

thaJeztah commented on api/swagger.yaml:
```diff
@@ -529,17 +529,6 @@ definitions:
         type: "array"
         items:
           $ref: "#/definitions/DeviceRequest"
-      KernelMemory:
```
can you also update the docs for the update endpoint in this file?

Commit:
Remove `KernelMemory` field from endpoints docs

Signed-off-by: aiordache <anca.iordache@docker.com>

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
@tianon ptal

tianon commented on daemon/daemon_unix.go:
```diff
@@ -463,17 +463,16 @@ func verifyPlatformContainerResources(resources *containertypes.Resources, sysIn
-	}
-	if resources.KernelMemory > 0 && !kernel.CheckKernelVersion(4, 0, 0) {
-		warnings = append(warnings, "You specified a kernel memory limit on a kernel older than 4.0. Kernel memory limits are experimental on older kernels, it won't work as expected and can cause your system to be unstable.")
+		if !sysInfo.KernelMemory {
```
Isn't the change over in `pkg/sysinfo/sysinfo_linux.go` deleting the code that sets this value?  (So this will never be invoked?)

thaJeztah replied:
Ahm, yes, you're right; I mistook that one for the `/info` endpoint (which _uses_ that information)

I guess instead, the systemInfo() endpoint would have to be updated to reset the field in the response for API >= 1.42 (and the field to have `omitempty`) https://github.com/moby/moby/blob/0fdd23b7fe4df100310ec5cac32dcb05e77dfcc2/api/server/router/system/system_routes.go#L46-L47

tianon reviewed (commented):
Given `runc` is now ignoring this value, I'm not sure I understand why we're working so hard to pretend it really is still supported on the older API levels?  Wouldn't it be more useful for users to bubble up that it's not going to do anything, even retroactively?

thaJeztah commented:
> Given runc is now ignoring this value, I'm not sure I understand why we're working so hard to pretend it really is still supported on the older API levels? Wouldn't it be more useful for users to bubble up that it's not going to do anything, even retroactively?

The main train of thought here was that;

- we (at least so far tried to) don't change existing APIs
- runtimes other than `runc` may still support it (although admittedly, they'd be living on borrowed time as the kernel removed support for this)

tianon reviewed (approved):
Discussed in maintainers meeting -- approach seems fair enough to me. :+1:

thaJeztah commented on daemon/daemon_unix.go:
```diff
@@ -463,17 +463,16 @@ func verifyPlatformContainerResources(resources *containertypes.Resources, sysIn
+			warnings = append(warnings, "Your kernel does not support kernel memory limit capabilities or the cgroup is not mounted. Limitation discarded.")
+			resources.KernelMemory = 0
+		}
+		if resources.KernelMemory < linuxMinMemory {
```
Looks like this should check if `resources.KernelMemory > 0`, otherwise this will produce an error if it's reset above (line 468)

thaJeztah commented:
For those arriving here; this was carried, and merged through https://github.com/moby/moby/pull/43214
